internal/pkg/api/handleDownload.go

```diff
@@ -0,0 +1,387 @@
+	// An additional check must be executed to validate it is not a random api key.
+	// This check is sufficient for the purposes of this API
+	_, err := authAPIKey(r, d.bulker, d.cache)
+	if err != nil {
+		return err
+	}
+
+	// check if file exists
+	if err := d.packageExists(requestPath); err != nil {
+		//get packages
+		if err := d.getPackage(requestPath); err != nil {
+			return err // return 503
+		}
+	}
+
+	// Wrap the response writer with the limiter and serve
+	wr := limit.WrapResponseWriter(d.ctx, w, d.bandwidthLimit)
+	h := http.StripPrefix("/api/downloads", http.FileServer(http.FS(d.fs)))
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	h.ServeHTTP(wr, r)
+	return nil
+}
+
+// packageExists will return an error if the requested package does not exist in the cache, or if the cached package has a size of 0
+func (d *Downloader) packageExists(requestPath string) error {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	// fs.Stat will reject any attempts to stat a path with ..
+	// this is good as it prevents requests attempting to gather non-cached files
+	fi, err := fs.Stat(d.fs, requestPath)
+	if err != nil {
+		return err
+	}
+	if fi.Size() < 1 {
+		return fmt.Errorf("package %q has length 0", requestPath)
+	}
+	return nil
+}
+
+// getPackage will cache the requested package using the upstreamURI as a source.
+// if the requested package is not an ".asc" or ".sha512" file then the associated files are downloaded and the package is verified.
+// All files are removed if an error is encountered.
+func (d *Downloader) getPackage(requestPath string) error {
+	// sanity check file path
+	filePath := filepath.Clean(filepath.Join(d.root, requestPath))
+	if !strings.HasPrefix(filePath, d.root) {
+		return fmt.Errorf("filepath invalid %q is not in package cache %q", filePath, d.root)
+	}
+
+	url, err := d.upstreamURI.Parse(path.Join("downloads", requestPath))
+	if err != nil {
+		return fmt.Errorf("unable to build url from %q: %w", requestPath, err)
+	}
+
+	err = d.getFile(url.String(), filePath)
+	if err != nil {
+		return err
+	}
+
+	// verify download (using sha and asc)
+	if !strings.HasSuffix(requestPath, ".asc") && !strings.HasSuffix(requestPath, ".sha512") {
+		err = d.getFile(url.String()+".asc", filePath+".asc")
+		if err != nil {
+			if rErr := d.removeFile(filePath); rErr != nil {
+				log.Error().Err(rErr).Msg("unable to remove file")
+			}
+			return err
+		}
+		err = d.getFile(url.String()+".sha512", filePath+".sha512")
+		if err != nil {
+			if rErr := d.removeFile(filePath); rErr != nil {
+				log.Error().Err(rErr).Msg("unable to remove file")
+			}
+			return err
+		}
+		err = d.verifier.Verify(filePath)
+		if err != nil {
+			if rErr := d.removeFile(filePath); rErr != nil {
+				log.Error().Err(rErr).Msg("unable to remove file")
+			}
+			return fmt.Errorf("package verification failed: %w", err)
+		}
+	}
+
+	// delete package later
+	if d.retention > 0 {
+		go d.startDeletionTimer(d.retention, filePath)
```

Comment:
this would cause multiple go routines scheduled to do deletions depending on timing, right? potentially as many as the getPackage calls?
maybe the better approach would be to expire the files on the next access, check the mod time?
and possible one go routine periodically managing expiration cleanups?